storage: add LOAD GENERATOR KEY VALUE

guswynn opened the pull request:
Adds a `LOAD GENERATOR` that goes through the upsert envelope.

This source is wildly easier to user than populating a kafka topic, and lets us control things more accurately. The sources let you configure the key-space, the number of partitions, and the value size. It also lets you configure the update rate and snapshot size, as well as emulate the pre-correctness-property-#2 kafka source with `QUICK ROUNDS`.


The intention of this source is to test upsert rehydration performance, as well as hydration backpressure. I added docs and a feature flag, but defaulted it to on. 

### Motivation

  * This PR adds a known-desirable feature.


### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  related to hydration backpressure (for testing purposes)
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

rjobanp commented on doc/user/content/sql/create-source/load-generator.md:
```diff
@@ -168,6 +176,34 @@ The TPCH source must be used with `FOR ALL TABLES`, which will create the standa
+- `QUICK ROUNDS`: The number of rounds of data (1 update per key in each round) to produce after the snapshot, but
+    before considering `UPDATE RATE`.
+- `VALUE SIZE`: The number of bytes in each `value.
+- `UPDATE RATE`: The _maximum_ rate to produce batches of data (within each partition) after snapshotting.
```
what unit is this?

guswynn replied:
its an interval, so it can be any unit!

rjobanp commented on doc/user/content/sql/create-source/load-generator.md:
```diff
@@ -168,6 +176,34 @@ The TPCH source must be used with `FOR ALL TABLES`, which will create the standa
+    before considering `UPDATE RATE`.
+- `VALUE SIZE`: The number of bytes in each `value.
+- `UPDATE RATE`: The _maximum_ rate to produce batches of data (within each partition) after snapshotting.
+- `SEED`: A per-source seed for seeding the random data.
```
value example?

rjobanp commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1239,17 +1266,33 @@ pub fn plan_create_source(
+                    ..
+                }) = &external_connection
+                {
+                    KeyEnvelope::Named("key".to_string())
```
nit - move `"key"` to a constant declared in `storage-types/src/sources/load_generator.rs`

rjobanp commented on src/storage/src/source/generator/upsert.rs:
```diff
@@ -0,0 +1,556 @@
+                    .await;
+            };
+
+            // Downgrade to the
```
finish comment

rjobanp commented on src/storage/src/source/generator/upsert.rs:
```diff
@@ -0,0 +1,556 @@
+        )>,
+        value_buffer: &mut Vec<u8>,
+    ) -> u64 {
+        // Consistently seeded with the source see, the offset, and the
```
```suggestion
        // Consistently seeded with the source seed, the offset, and the
```

rjobanp commented on src/storage/src/source/generator/upsert.rs:
```diff
@@ -0,0 +1,556 @@
+            while local_partitions.values().any(|si| !si.finished()) {
+                for sp in local_partitions.values_mut() {
+                    updates_buffer.clear();
+                    emitted += sp.produce_batch(&mut updates_buffer, &mut value_buffer);
```
if we produced 0 we could early continue here, but just a nit

guswynn replied:
I think its probably clearer this way!

rjobanp commented on src/storage/src/source/generator/upsert.rs:
```diff
@@ -0,0 +1,556 @@
+                        )
+                    })
+                })
+                .collect();
```
if we have no `local_partitions` we can early return, right? otherwise it looks like we might hot-loop below (the `yield_now` is preventing that but it might be unnecessary)

guswynn replied:
we will loop on the `interval.tick()`, but I can break out I think!

rjobanp commented on src/storage/src/source/generator/upsert.rs:
```diff
@@ -0,0 +1,556 @@
+        }
+    }
+
+    /// If this partition is done snapshotting.
```
```suggestion
    /// If this partition is done producing quick updates.
```

rjobanp commented on src/storage/src/source/generator/upsert.rs:
```diff
@@ -0,0 +1,556 @@
+        batch_size: u64,
+        seed: u64,
+        next_offset: u64,
+    ) -> Self {
+        // The non-snapshot offset * the number of keys in a batch partitioned across the key
+        // space is the _key_ of the next batch at `next_offset`.
+        let start_key =
+            (((next_offset - snapshot_rounds) * batch_size * partitions) + partition) % keys;
+
+        // We expect to emit 1 batch per offset.
+        let expected_quick_offsets =
+            ((keys / partitions / batch_size) * quick_rounds) + snapshot_rounds;
+
```
this is a little hard to wrap my head around, and I wonder if it's because `round` and `offset` are being used for the same concept? Is that correct or am I misunderstanding?

guswynn replied:
They are very very similar, I will clarify!

rjobanp reviewed (commented):
nice work!

guswynn commented:
Thanks for the nits @rjobanp, will get to them!

Offline @petrosagg had one more suggestion: don't just assume `ENVELOPE UPSERT`, instead require an `ENVELOPE` parameter. This makes some things less special cased, and also lets use this load gen in a more general purpose way. I'll make sure to add that as a separate commit

benesch commented:
Thought from the SQL council perspective: is `UPSERT` the right name for this load gen? It doesn't have to be used with `ENVELOPE UPSERT`, right? I think `LOAD GENERATOR KEY VALUE` might be a better name, because it doesn't prescribe that it must be used with `ENVELOPE UPSERT`!

benesch commented on src/sql/src/session/vars/definitions.rs:
```diff
@@ -2006,6 +2006,13 @@ feature_flags!(
+    {
+        name: enable_load_generator_upsert,
+        desc: "Create a LOAD GENERATOR UPSERT",
+        default: true,
```
Insta-stabilizing this seems a bit scary? Since the primary need right now is for internal testing, seems safer to default `false` and stabilize it after we have some experience with it?

mjibson commented on src/sql-parser/tests/testdata/ddl:
```diff
@@ -2361,6 +2361,13 @@ CREATE SOURCE lg FROM LOAD GENERATOR COUNTER
 CreateSource(CreateSourceStatement { name: UnresolvedItemName([Ident("lg")]), in_cluster: None, col_names: [], connection: LoadGenerator { generator: Counter, options: [] }, include_metadata: [], format: None, envelope: None, if_not_exists: false, key_constraint: None, with_options: [], referenced_subsources: None, progress_subsource: None })
 
+parse-statement
+CREATE SOURCE lg FROM LOAD GENERATOR UPSERT (KEYS 1, PARTITIONS 2, UPDATE RATE '1m', BATCH SIZE 100, SEED 200, VALUE SIZE 150)
```
Can you add all new options here?

guswynn replied:
oh yeah, forgot to add em!

mjibson reviewed (commented):
Adapter parts lgtm.

guswynn commented:
Ok, I think I got everyone's review:
- @ParkMyCar @benesch @petrosagg : upsert and none envelopes are supported
  - as part of this I also added `INCLUDE OFFSET` support
- Fixed up docs and also roshan and matt's suggestions

petrosagg commented on src/storage-types/src/sources/load_generator.rs:
```diff
@@ -52,7 +57,14 @@ impl SourceConnection for LoadGeneratorSourceConnection {
 
     fn key_desc(&self) -> RelationDesc {
-        RelationDesc::empty()
+        if let LoadGenerator::KeyValue(_) = &self.load_generator {
```
Can we make this a match?

petrosagg commented on src/storage-types/src/sources/load_generator.rs:
```diff
@@ -339,7 +362,6 @@ impl LoadGenerator {
                             .with_column("l_shipdate", ScalarType::Date.nullable(false))
                             .with_column("l_commitdate", ScalarType::Date.nullable(false))
                             .with_column("l_receiptdate", ScalarType::Date.nullable(false))
-                            .with_column("l_shipinstruct", ScalarType::String.nullable(false))
```
what happened here? Is this a bug?

guswynn replied:
omg yes, VIMMMMM

petrosagg commented on src/storage-types/src/sources/load_generator.rs:
```diff
@@ -376,6 +399,7 @@ impl LoadGenerator {
             LoadGenerator::Marketing => false,
             LoadGenerator::Datums => true,
             LoadGenerator::Tpch { .. } => false,
+            LoadGenerator::KeyValue(_) => false,
```
The key value generator is monotonic, no?

guswynn replied:
ah yes!

petrosagg commented on src/storage-types/src/sources/load_generator.rs:
```diff
@@ -447,8 +472,68 @@ impl RustType<ProtoLoadGeneratorSourceConnection> for LoadGeneratorSourceConnect
+
+#[derive(Arbitrary, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Default)]
+pub struct KeyValueLoadGenerator {
+    // The keyspace of the source.
```
All of these should be doc comments
```suggestion
    /// The keyspace of the source.
```

petrosagg commented on src/storage-types/src/sources/load_generator.rs:
```diff
@@ -447,8 +472,68 @@ impl RustType<ProtoLoadGeneratorSourceConnection> for LoadGeneratorSourceConnect
+            quick_rounds: self.quick_rounds,
+            value_size: self.value_size,
+            partitions: self.partitions,
+            update_rate: self.update_rate.as_ref().map(|ur| ur.into_proto()),
```
We have generic implementations for this

```suggestion
            update_rate: self.update_rate.into_proto(),
```

guswynn replied:
oh nice!

petrosagg commented on src/storage-types/src/sources/load_generator.rs:
```diff
@@ -447,8 +472,68 @@ impl RustType<ProtoLoadGeneratorSourceConnection> for LoadGeneratorSourceConnect
+            quick_rounds: proto.quick_rounds,
+            value_size: proto.value_size,
+            partitions: proto.partitions,
+            update_rate: proto.update_rate.map(|ur| ur.into_rust()).transpose()?,
```
Same here 
```suggestion
            update_rate: proto.update_rate.into_rust()?,
```

petrosagg commented on src/storage/src/render/sources.rs:
```diff
@@ -203,9 +203,9 @@ where
+                key: Some(Ok(r.key)),
                 value: Some(Ok(r.value)),
-                metadata: Row::default(),
+                metadata: r.metadata,
```
nice

guswynn replied:
the gud stuff

petrosagg commented on src/storage/src/source/generator.rs:
```diff
@@ -63,6 +64,7 @@ pub fn as_generator(g: &LoadGenerator, tick_micros: Option<u64>) -> Box<dyn Gene
             // completely.
             tick: Duration::from_micros(tick_micros.unwrap_or(0)),
         }),
+        LoadGenerator::KeyValue(KeyValueLoadGenerator { .. }) => panic!("not a basic generator"),
```
what is a basic generator? Are we missing some enum somewhere?

petrosagg commented on src/storage/src/source/generator.rs:
```diff
@@ -139,7 +147,7 @@ impl SourceRender for LoadGeneratorSourceConnection {
                             output,
                             Ok(SourceMessage {
-                                key: Row::default(),
+                                key: Row::pack_slice(&[Datum::UInt64(1)]),
```
Why did we change the schema of the other generators?

guswynn replied:
oh this was a testing thing. I think in general we drop this key but it may be the flake? not sure

petrosagg commented on src/storage/src/source/generator.rs:
```diff
@@ -76,14 +78,20 @@ impl SourceRender for LoadGeneratorSourceConnection {
     ) {
+        // Currently `KEY VALUE` loadgen renders its own operator, and is not a single-worker
+        // basic generator like the rest.
+        if let LoadGenerator::KeyValue(kv) = self.load_generator.clone() {
```
I think we need an enum of `GeneratorKind { Simple(..), KeyValue(..) }` where each of the inner types has a render method.

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+    key_value: KeyValueLoadGenerator,
+    scope: &mut G,
+    config: RawSourceCreationConfig,
+    resume_uppers: impl futures::Stream<Item = Antichain<MzOffset>> + 'static,
```
Let's start using this new name
```suggestion
    committed_upper: impl futures::Stream<Item = Antichain<MzOffset>> + 'static,
```

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+                        p,
+                        SnapshotProducer::new(
+                            p,
+                            key_value.partitions,
```
Should we just pass `&key_value` here?

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+                        p,
+                        UpdateProducer::new(
+                            p,
+                            key_value.partitions,
```
Should we just pass `&key_value` here?

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+
+        let snapshotting = resume_offset.offset == 0;
+
+        let mut local_partitions: BTreeMap<_, _> = (0..key_value.partitions)
```
Why is this a map? The keys are never read

guswynn replied:
No real reason, other than the fact that I don't need to reason about the `Ord` impl of the generator structs

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+            resume_offset.offset
+        };
+
+        let mut local_partitions: BTreeMap<_, _> = (0..key_value.partitions)
```
Why is this a map? The keys are never read

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+
+/// An iterator that produces keys belonging to a partition.
+struct PartitionKeyIterator {
+    // The partition.
```
Doc comments for the fields 
```suggestion
    /// The partition.
```

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ret = self.next;
+        self.next = (self.next + self.partitions) % self.keys;
```
This is only correct with the assumption of partitions being a divisor of keys. I think what you need can be expressed more robustly with this:

```rust
let keys = (0..keys).filter(|k| (k % partitions) == partition).cycle();
```

guswynn replied:
I assert this assumption in `new`, don't I?

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+
+/// A struct that produces batches of data for the snapshotting phase.
+struct SnapshotProducer {
+    // The key iterator for the partition.
```
doc comments here too 
```suggestion
    /// The key iterator for the partition.
```

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+
+/// A struct that produces batches of data for the post-snapshotting phase.
+struct UpdateProducer {
+    // The key iterator for the partition.
```
```suggestion
    /// The key iterator for the partition.
```

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+        )>,
+        value_buffer: &mut Vec<u8>,
+    ) -> u64 {
+        // Consistently seeded with the source see, the offset, and the
```
```suggestion
        // Consistently seeded with the source seed, the offset, and the
```

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+        seed[0..8].copy_from_slice(&self.seed.to_le_bytes());
+        seed[8..16].copy_from_slice(&self.next_offset.to_le_bytes());
+        seed[16..24].copy_from_slice(&self.pi.partition.to_le_bytes());
+        let mut rng = StdRng::from_seed(seed);
```
This should be extracted in its own function since it's copy pasted in two places currently and it's critical for the source to produce a sensible collection

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+pub fn render_statistics_operator<G: Scope<Timestamp = MzOffset>>(
+    scope: &mut G,
+    config: RawSourceCreationConfig,
+    resume_uppers: impl futures::Stream<Item = Antichain<MzOffset>> + 'static,
```
```suggestion
    comitted_upper: impl futures::Stream<Item = Antichain<MzOffset>> + 'static,
```

petrosagg commented on src/storage/src/source/generator/key_value.rs:
```diff
@@ -0,0 +1,584 @@
+            0,     // not important
+            2,     // batch size 2
+            1234,  //seed
+            5,     // resume upper of 5
```
Yeah.. we should definitely be passing the struct with the named fields here

guswynn replied:
yeah I agree!

petrosagg commented on src/storage/src/source/source_reader_pipeline.rs:
```diff
@@ -258,7 +258,7 @@ fn source_render_operator<G, C>(
     config: RawSourceCreationConfig,
     source_connection: C,
-    resume_uppers: impl futures::Stream<Item = Antichain<C::Time>> + 'static,
+    resume_uppers: impl futures::Stream<Item = Antichain<C::Time>> + Send + 'static,
```
What needed this?

guswynn replied:
oh this was from something previous, will remove

guswynn commented on doc/user/content/sql/create-source/load-generator.md:
```diff
@@ -168,6 +176,37 @@ The TPCH source must be used with `FOR ALL TABLES`, which will create the standa
+
+The following options are supported:
+
+- `KEYS`: The number of keys in the source. Must be divisible by `PARTITIONS` * `BATCH SIZE`
```
It makes the math way easier, no dealing with the extra dangling bit

we can extend it later if need be

guswynn commented on doc/user/content/sql/create-source/load-generator.md:
```diff
@@ -168,6 +176,37 @@ The TPCH source must be used with `FOR ALL TABLES`, which will create the standa
+- `QUICK ROUNDS`: The number of rounds of data (1 update per key in each round) to produce after the snapshot, but
+    before considering `UPDATE RATE`.
+- `VALUE SIZE`: The number of bytes in each `value.
+- `UPDATE RATE`: The _maximum_ rate (as an [`interval`]) to produce batches of data (within each partition) after snapshotting.
```
good point! Will use `TICK INTERVAl`

guswynn commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1223,10 +1230,24 @@ pub fn plan_create_source(
             (key_desc, value_desc)
         }
-        None => (None, external_connection.value_desc()),
+        None => (
```
How did this even happen??

guswynn commented on doc/user/content/sql/create-source/load-generator.md:
```diff
@@ -168,6 +176,37 @@ The TPCH source must be used with `FOR ALL TABLES`, which will create the standa
+- `SNAPSHOT ROUNDS`: The number of rounds of data (1 update per key in each round) to produce
+    as the source starts up. Can be used to scale the size of the snapshot without changing the number
+    of keys.
+- `QUICK ROUNDS`: The number of rounds of data (1 update per key in each round) to produce after the snapshot, but
```
changed to `TRANSACTIONAL SNAPSHOT` as discussed offline

guswynn commented:
@petrosagg I believe I have covered all your review. There is still one special case in planning:

`KEY VALUE` load gens get to have a key envelope even if they have no encoding. I believe this all goes away when we turn planning into stages, and decoupling the decode and envelope stages. It seems we currently have some emergent support for decoding non-kafka sources, but its all a mess in `ddl.rs`, and we need to come clean it up. I was able to clean up some small other things, based on your suggestions tho

petrosagg commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1223,10 +1230,30 @@ pub fn plan_create_source(
+    // TODO(guswynn|petrosagg): when this is refactored, remove this special-case
+    // and move the `bail_unsupported` call below to a check specific to planning
+    // the decode stage.
+    let key_envelope_no_encoding = matches!(
```
I don't think we need this. If you use the key value load generator without `ENVELOPE UPSERT` then you should explicitly specify `INCLUDE KEY` if you want to see the key. When you do specify `ENVELOPE UPSERT` then the planner already auto-sets `INCLUDE KEY` if not specified by the user

guswynn replied:
The code that does that panics; it relies on an `encoding` being available :(

guswynn replied:
I looked at this closer today:

Note that the code as it stands does the behavior you specify: `INCLUDE KEY` is not auto-set, unless its specified by the user OR they use `ENVELOPE UPSERT`. This is tested at the bottom on the td file in the pr.

Removing this special case boils down to removing these errors:
```
> INCLUDE KEY requires specifying KEY FORMAT .. VALUE FORMAT, got bare FORMAT`
> UPSERT requires a key/value format: {:?}
> ENVELOPE [DEBEZIUM] UPSERT requires that KEY FORMAT be specified
```

(These are actually all very very similar errors. There is a world where we invert the special case, and much earlier in planning assert that Kafka sources have the above requirements, but thats still leaving a special case in), which amounts to deciding to:

  default the key envelope of kafka sources to BYTES
  
Is that something we want to commit to?

petrosagg commented on doc/user/content/sql/create-source/load-generator.md:
```diff
@@ -168,6 +175,37 @@ The TPCH source must be used with `FOR ALL TABLES`, which will create the standa
+- `SNAPSHOT ROUNDS`: The number of rounds of data (1 update per key in each round) to produce
+    as the source starts up. Can be used to scale the size of the snapshot without changing the number
+    of keys.
+- `TRANSACTIONAL SNAPSHOT`: Whether or not to emit the snapshot as a singular transaction.
```
Nice!

guswynn replied:
much clearer!

petrosagg commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1710,6 +1766,82 @@ pub(crate) fn load_generator_ast_to_generator(
+                partitions,
+                batch_size,
+                ..
+            } = options.to_vec().try_into()?;
```
There is a variable named `extracted` in scope with this conversion and it also exists in the caller of this function in line 1082. Let's convert the options once somewhere in `1082` and pass a `&LoadGeneratorOptionExtracted` in this funciton so that we don't do the conversion over and over.

There is also a call to `ensure_only_valid_options` here that should be hoisted

guswynn replied:
Oh this is because of a bad rebase, will fixup

guswynn commented:
@petrosagg Updated the pr with additional context based on our discussion, also documented here: https://github.com/MaterializeInc/materialize/pull/26246#issuecomment-2023558097

I think everything else is resolved and we can merge this!

petrosagg commented:
@guswynn thank you and sorry for the ddl.rs sidetrack!

guswynn commented:
@petrosagg no worries, it was super useful to figure out!
